### DISCUSSION TITLE
Implement custom versioning logic???

### DISCUSSION THREAD

**Discussion by poloka** [lerna/lerna#3681] [Q&A]
Can I implement custom versioning logic during 'lerna version' and 'lerna publish' stages?

I have a use case where I want to publish a prerelease with a version that looks like the following:
* `3.0.0-<preid>.0`

The current lerna publish seems to add a `sha` reference at the end which I do not want.  What I do want is to increment the last identifier so I can get
* `3.0.0-<preid>.0`
* `3.0.0-<preid>.1`
* `3.0.0-<preid>.2`

I tried tying into the post processing lifecycle event `version` in my package.json to execute a script to do my own custom modifications but upon saving my updates it is only reflected in my package.json.  Any project that has a reference to this project in my monorepo fails to find the updated version.  The standard `lerna publish` updated the version to `3.0.0-foobar.0.1812`.  My custom script knocks off the `1812` and updates the version to be pushed to `3.0.0-foobar.1` because I already have a `3.0.0-foobar.0` pushed to my repository.  Now because I did this, during this build I have another project in my monorepo that was to be updated with the new version but I get an error saying 
```
lerna ERR! npm ERR! code ETARGET
lerna ERR! npm ERR! notarget No matching version found for MyProject@^3.0.0-foobar.0.1812.
```
It still has the original version stored in its processing that it want to update projects with.  

So want to know from the community what has been done and what is possible.  Is there an argument to not include the sha?  Can I change the version being utilize for this artifact stored within the current processing of the lerna command?  Can I replace the version logic utilize during the `lerna version` execution or do I only have access to pre and post actions?  Just want to see what my options are.

**Comment by fahslaj** [comment]
Hi @poloka , thank you for your question. I would only expect the git sha to be added if using the `--canary` option with `lerna publish`. If you run lerna and pass the positional argument of `prerelease` to `lerna version` or `lerna publish`, I would expect the behavior that you have described. Can you please elaborate on what commands you are running and what configuration you have in `lerna.json`?

**Comment by poloka** [comment]
Hey @fahslaj , thanks for the quick response.  Okay, need to add a bit more context.  I have a CI build that is pushing snapshots to development repository so I've been using --canary; otherwise, my CI is stuck waiting for the `Select a new version` prompt.  The command that was being generated and executed by my CI would look something like:
```
 npx lerna publish --registry <my-dev-repo> --canary --preid <branch-name> --tag-version-prefix <branch-name> --conventional-prerelease --yes
```
My `lerna.json` is:
```json
{
  "version": "independent",
  "useWorkspaces": true
}
```
Pretty basic at this point.  :-)

**Reply by fahslaj** [reply]
Got it! The `--canary` option will specifically add the git sha to the end of the commit. The "select a new version" prompt is given when a positional `bump` argument is not provided. Try replacing `--canary` with the word `prerelease` and you can get a version number like what you requested in the original question. See the [bump](https://github.com/lerna/lerna/tree/main/libs/commands/version#semver-bump) argument for other options besides "prerelease".

**Comment by poloka** [comment]
Cool, that helps.  Now I'm in a situation where I want to start with say 3.0.0.  By adding the `lerna publish prerelease` I keep getting a `3.0.1-<preid>.x`.  Is there a way to start with base version of `3.0.0`?

**Reply by fahslaj** [reply]
If your packages are already at version 3.0.0, then the next prerelease version is 3.0.1-<preid>.x because according to semver, 3.0.0 is a later version than 3.0.0-<preid>.x. If you want Lerna to pick 3.0.0 as a version, then you will need to start from a version before that. There are also other options than prerelease, such as `premajor`, `preminor`, and `prepatch`. You can see the full list of those options [here in the lerna version docs](https://github.com/lerna/lerna/tree/main/libs/commands/version#semver-bump)

**Comment by poloka** [comment]
Also, if I was running this prepublish from my CI, how can subsequent builds know that the version goes from `3.0.0-<preid>.0` to `3.0.0-<preid>.1`?  Do I have to have the CI commit the updates? 

Previously, what I was planning on doing was to pull the results from 
```
npm view --registry <registry> --json <project-name> versions
```
And then identify the latest version and increment the version from the `0` to a `1`.

**Comment by poloka** [comment]
Thought of another question, if I was to write some custom logic for doing stuff on a prerelease, would there be a way to tell this execution path version a standard release?  I want to run something in the `version` lifecycle but only for prerelease.  I can't add the `--ignore-scripts` tag as I may have other items that I'd like to do when publishing on the post version updates.

**Comment by poloka** [comment]
Where I am stuck is that I need to tell lerna the version has been updated.  I tried in the `preversion` and also logic in the `version` lifecylce event.  But regardless of my updates to the 'version' attribute in the `package.json`, lerna looks to have stored the version and does not let go of what it thinks is the correct version without reevaluating the `package.json` after these lifecycle events have completed.
```
Changes:
 - FooBar: 3.0.0 => 3.0.1-addsnapshotsave.0

lerna info auto-confirmed 
lerna info execute Skipping git tag/commit
lerna info execute Skipping git push
lerna info execute Skipping releases
lerna info lifecycle script-parent@1.0.0~preversion: foobar@1.0.0

> foobar@1.0.0 preversion /Users/me/foobar
> node ./scripts/lerna-preversion-lifecycle.js
```
Is there anything that can be done?  I'm stuck and looks like lerna is not very flexible in this aspect.

**Reply by fahslaj** [reply]
Lerna will use the version number in `lerna.json` to determine the next version, so if you are manually changing the version beforehand anywhere, you will want to update that as well.

**Reply by poloka** [reply]
Hmm, I don't see any updates to my `lerna.json` file when running the `lerna version` command.   I'm not seeing how this file is used to store the versions associated to the packages being incremented.

**Reply by poloka** [reply]
I even tried writing my next version in a `lerna.json` file local to each project so I have one in my FooBar project that looks like:
```json
{
  "version": "3.0.1-addsnapshotsave.1"
}
```
And it didn't make any difference in the version updating.  Still trying to publish `3.0.1-addsnapshotsave.0`

**Reply by poloka** [reply]
Since my packages are independent, I was writing to the `packages/<project>/lerna.json`.  That didn't work.  I attempted to add a version number to my root `lerna.json`, made a change, but then lerna writes back a config during its version update implementation.  Unless I'm missing something, I can't seem to update the version utilized because lerna looks to be holding onto a version number in its implementation and not taking into account that things could be updated in either the pre version process or post process when going to publish.

**Reply by fahslaj** [reply]
There should only be one `lerna.json` file, and it should be in the root. The option `--conventional-prerelease` that is specified in your command above will tell Lerna to suggest a new version based on the [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/) specification. You can try removing that option, since it sounds like you want to control what version is picked.

If "version" in lerna.json is equal to "independent", then Lerna will use the version in each package's package.json file as the base on which to pick a new version from. If the version in lerna.json is an actual version, then it will use that for all packages.

So if you have "version": "independent" in lerna.json in the root, then change a single package with its version set to "3.0.0" in its package.json, running `npm run version prerelease` will result in the version "3.0.1-<preid>.0 for that package.
